src/game_api/script.cpp

```diff
@@ -1037,6 +1038,27 @@ SpelunkyScript::ScriptImpl::ScriptImpl(std::string script, std::string file, Sou
         }
         return (uint16_t)0;
     };
+    /// Read input that has been previously stolen with steal_input
+    lua["read_stolen_input"] = [this](int uid)
+    {
+        Player* player = get_entity_ptr(uid)->as<Player>();
+        if (player == nullptr)
+            return (uint16_t)0;
+        ScriptInput* readinput = reinterpret_cast<ScriptInput*>(player->input_ptr);
+        if (!IsBadReadPtr(readinput, 20))
+        {
+            if (!readinput->is_input_stolen)
```

Comment:
So normally the game points to just a 32bit large struct, let's call it `GameInput`
```
struct GameInput {
    uint16_t next;
    uint16_t current;
};
```
Now when stealing the input OL allocates `ScriptInput` which shares a common initial sequence (aka the first two members have same alignment and size) And OL can access this pointer as a `ScriptInput` if and only if it has allocated it as a `ScriptInput`. What you are doing is to assume it always contains a `ScriptInput` when in reality it could contain a `GameInput`. So you might read past the end of the struct and read trash data. I know the intention is to only use this function when the input is stolen, but that's no good reason to assume it will only be called in that case.
So what you really need to do is check if `this.script_input` contains `uid`, if it does you treat the input as stolen, otherwise you treat it as vanilla.

Reply:
I think I fixed this? Can you take another look?